### PR TITLE
MAINT: Use accessor function for numpy array internals

### PR DESCRIPTION
**Pull request by jaimefrio**
These PR replaces all accesses in ndimage to the internal elements of a numpy array struct by calls to the relevant accessor functions.

Trying to use the accessor functions unveiled some issues with the use of the numpy API across compilation units, see numpy/numpy#9309 for some context, and the notes in `nd_image.c` and `ni_support.h` for the mechanism now in place.

This last change also forced the move of some functions from `nd_image.h` to `nd_image.c`.

There are also fixes to the format on lines touched by the above changes, and the usual trailing whitespace removal due to opening the files with a proper text editor...

### CODE REVIEW

File: scipy/ndimage/src/nd_image.h [hunk 1]
```diff
@@ -34,12 +34,8 @@
 
 #include "Python.h"
 
-#ifndef ND_IMPORT_ARRAY
-#define NO_IMPORT_ARRAY
-#endif
-
+#define PY_ARRAY_UNIQUE_SYMBOL _scipy_ndimage_ARRAY_API
```
Comment (perimosocordiae) [inline scipy/ndimage/src/nd_image.h]: Might be a good idea to leave a comment explaining how this works, maybe linking to https://docs.scipy.org/doc/numpy/reference/c-api.array.html#importing-the-api.

**Review by perimosocordiae — APPROVED** [review]
This is a pretty big diff, and I have to admit I haven't read through the entire thing to make sure all the direct member accesses are now accessor macros, but overall it looks good to me.

File: scipy/ndimage/src/nd_image.h (hunk 1 above)
Comment (jaimefrio) [inline scipy/ndimage/src/nd_image.h]: So you found what google couldn't, see numpy/numpy#9309...

**Comment by jaimefrio** [comment]
Yes, apologies for the huge PR. If it's too much to swallow I can break it down into half a dozen smaller ones, just ask!

Will certainly do for other similar changes that affect many files.

**Comment by rgommers** [comment]
> These PR replaces all accesses in ndimage to the internal elements of a numpy array struct by calls to the relevant accessor functions.

There's still a bunch left, like:

    scipy/ndimage/src/nd_image.c:1029:66: error: ‘PyArrayObject’ has no member named ‘nd’
                 end = PyLong_FromSsize_t(regions[idx + jj + input->nd]);

Output from build with deprecated API disabled. Note that that commit was intended for inclusion in a PR like this, not just for testing. Or did you leave it out on purpose?

**Comment by jaimefrio** [comment]
I had missed those ones...

Have added your build changes in the last commit, and fixed a number of macro names and added some pointer casts to silence warnings.

**Comment by rgommers** [comment]
LGTM now, merging. Thanks @jaimefrio @perimosocordiae

**Pull request merged**